openstudiocore/src/model/EnergyManagementSystemProgram.cpp

```diff
@@ -0,0 +1,430 @@
+#include "Model.hpp"
+
+#include "../utilities/core/String.hpp"
+#include "../utilities/core/StringHelpers.hpp"
+#include "../utilities/core/UUID.hpp"
+#include <boost/algorithm/string.hpp>
+
+//#include "../utilities/idf/Handle.hpp"
+
+namespace openstudio {
+namespace model {
+
+namespace detail {
+
+  EnergyManagementSystemProgram_Impl::EnergyManagementSystemProgram_Impl(const IdfObject& idfObject,
+                                                                         Model_Impl* model,
+                                                                         bool keepHandle)
+    : ModelObject_Impl(idfObject,model,keepHandle)
+  {
+    OS_ASSERT(idfObject.iddObject().type() == EnergyManagementSystemProgram::iddObjectType());
+  }
+
+  EnergyManagementSystemProgram_Impl::EnergyManagementSystemProgram_Impl(const openstudio::detail::WorkspaceObject_Impl& other,
+                                                                         Model_Impl* model,
+                                                                         bool keepHandle)
+    : ModelObject_Impl(other,model,keepHandle)
+  {
+    OS_ASSERT(other.iddObject().type() == EnergyManagementSystemProgram::iddObjectType());
+  }
+
+  EnergyManagementSystemProgram_Impl::EnergyManagementSystemProgram_Impl(const EnergyManagementSystemProgram_Impl& other,
+                                                                         Model_Impl* model,
+                                                                         bool keepHandle)
+    : ModelObject_Impl(other,model,keepHandle)
+  {}
+
+  const std::vector<std::string>& EnergyManagementSystemProgram_Impl::outputVariableNames() const
+  {
+    static std::vector<std::string> result;
+    if (result.empty()){
+    }
+    return result;
+  }
+
+  IddObjectType EnergyManagementSystemProgram_Impl::iddObjectType() const {
+    return EnergyManagementSystemProgram::iddObjectType();
+  }
+
+  boost::optional<std::string> EnergyManagementSystemProgram_Impl::body() const {
+    //return program body as string
+
+    // loop through extensible groups and add ProgramLine to body string.
+    std::string body;
+    boost::optional<std::string> comment;
+
+    auto groups = extensibleGroups();
+    for (auto group = groups.begin(); group != groups.end(); ++group) {
+      //get program line content
+      const auto & line = group->getString(OS_EnergyManagementSystem_ProgramExtensibleFields::ProgramLine, true);
+      OS_ASSERT(line);
+      body += line.get();
+      //get non-default comments if they exist
+      comment = group->fieldComment(OS_EnergyManagementSystem_ProgramExtensibleFields::ProgramLine, false);
+      if (comment.is_initialized()) {
+        //remove space after !
+        boost::erase_first(comment.get()," ");
+        //add space between end of program line and comment
+        if (!comment.get().empty()) {
+          body += " " + comment.get();
+        }
+      }
+      //add newline character
+      body += '\n';
```

Comment:
I've been developing on Windows and ran a converted PTool measure to use the new stuff and it seemed to work fine.